### DISCUSSION TITLE
Confusion about which device tree is used - does not appear to be `A_kernel-dtb` nor `B_kernel-dtb` on `scarthgap`

### DISCUSSION THREAD

**Discussion by tbottomparallel** [OE4T/meta-tegra#1670] [Q&A] [answered]
I am in the process of updating to `scarthgap` and am confused about what device tree is being used. We're now using A/B roofs and after flashing the `B_kernel-dtb` partition and rebooting to the B slot, the expected changes to the device tree were not present:

```
"dd" "status=progress" "bs=8M" "iflag=fullblock" "oflag=direct" "conv=sparse" "if=kernel_tegra234-p3737-0000+p3701-0000-nv.dtb" "of=/dev/disk/by-partlabel/B_kernel-dtb"
```

My initial suspicion was that the A slot device tree was still being used. In an attempt to demonstrate this I created a modified device tree for the A and B slots that would identify which slot was in use:

```console
$ dtc -I dtb -O dts A_kernel_modified.dtb 2>&1 | grep model
        model = "NVIDIA Jetson AGX Orin Developer Kit A";

$ dtc -I dtb -O dts B_kernel_modified.dtb 2>&1 | grep model
        model = "NVIDIA Jetson AGX Orin Developer Kit B";
```

I flashed the modified device trees to the `A_kernel-dtb` and `B_kernel-dtb` slots respectively:

```console
$ dd if=A_kernel_modified.dtb of=/dev/disk/by-partlabel/A_kernel-dtb
532+1 records in
532+1 records out
272497 bytes (272 kB, 266 KiB) copied, 0.0317612 s, 8.6 MB/s

$ dd if=B_kernel_modified.dtb of=/dev/disk/by-partlabel/B_kernel-dtb
531+1 records in
531+1 records out
272233 bytes (272 kB, 266 KiB) copied, 0.0302397 s, 9.0 MB/s
```

rebooted and observed that neither modified device tree appeared to be in use:

```console
$ cat /sys/firmware/devicetree/base/model
NVIDIA Jetson AGX Orin Developer Kit
```

I read the device trees back and verified that the changes were present on disk:

```console
$ dd if=/dev/disk/by-partlabel/A_kernel-dtb of=A_kernel-readback.dtb
1536+0 records in
1536+0 records out
786432 bytes (786 kB, 768 KiB) copied, 0.0104452 s, 75.3 MB/s

$ dtc -I dtb -O dts A_kernel-readback.dtb 2>&1 | grep model
        model = "NVIDIA Jetson AGX Orin Developer Kit A";

$ dd if=/dev/disk/by-partlabel/B_kernel-dtb of=B_kernel-readback.dtb
1536+0 records in
1536+0 records out
786432 bytes (786 kB, 768 KiB) copied, 0.0113336 s, 69.4 MB/s

$ dtc -I dtb -O dts B_kernel-readback.dtb 2>&1 | grep model
        model = "NVIDIA Jetson AGX Orin Developer Kit B";
```

Out of concern that a driver might be truncating or modifying the model field behind my back, I also tried adding a new node to the `B_kernel-dtb`, which likewise does not appear in the device tree:

```console
$ dtc -I dtb -O dts B_kernel_modified.dtb 2>&1 | grep -A5 new-test-node
        new-test-node {
                compatible = "no-driver-exists";
                status = "okay";
        };

$ nvbootctrl dump-slots-info
Current version: 36.3.0
Capsule update status: 1
Current bootloader slot: B
Active bootloader slot: B
num_slots: 2
slot: 0,             status: normal
slot: 1,             status: normal

$ ls /sys/firmware/devicetree/base/new-test-node
ls: cannot access '/sys/firmware/devicetree/base/new-test-node': No such file or directory
```

The device tree does not change regardless of what I write to these partitions. Prior to scarthgap we always just wrote the device tree to `A_kernel-dtb` and this worked without issue.

So my questions are:

**Comment by dwalkes** [comment, marked as answer]
See https://github.com/OE4T/meta-tegra/commit/d4dfb31b34601e31a387fd42dc79555f06194ab3, https://github.com/OE4T/meta-tegra/wiki/extlinux.conf-support and the logic in https://github.com/OE4T/meta-tegra/blob/scarthgap/recipes-bsp/uefi/l4t-launcher-extlinux.bb - the dtb partitions are only used for backup I believe as long as `UBOOT_EXTLINUX`, which is now the default on `master` and `scarthgap`, is used.

**Reply by tbottomparallel** [reply]
thanks for the quick reply!

All I needed was `UBOOT_EXTLINUX_FDT = "/boot/${DTBFILE}"`, now my devicetree gets updated along with the rootfs.

**Comment by johannesschrimpf** [comment]
I'm still wondering about the initial question:

**Where is the device tree stored if not in A_kernel-dtb nor B_kernel-dtb?**

While the answer might fix the specific problem, I still wonder where the actually kernel takes the device tree from if not from the rootfs and neither the boot partition.

In my setup, the dtb is not in `/boot`. I overwrite the kernel partitions with zeros, and still my system boots.

Is it also flashed into the boot loaders or somewhere else?



As for the second question:
**How can I configure it to use A_kernel-dtb and B_kernel-dtb?**

If I understand correctly, the `UBOOT_EXTLINUX_FDT` will install the dtb to `/boot` and will not use the boot partition.

As for how to use the partitions: At least for the Orin NX, you can enter the boot setup by pressing esc, then `Device Manager->NVIDIA Configuration->L4T Configuration` and set `L4T Boot Mode` from ExtLinux to Kernel Partition. Then it seems to use the partitions instead of the mysterious dtb location.

Still I wonder how I could set this up in the yocto build directly without going into the boot meny manually.

**Reply by madisongh** [reply]
> Is it also flashed into the boot loaders or somewhere else?

Yes, there's a copy of the device tree built into the UEFI bootloader, and if I remember correctly, that's used as a fallback if the extlinux and Android partition mechanisms don't locate the kernel device tree.

> Still I wonder how I could set this up in the yocto build directly without going into the boot meny manually.

There might be a way to configure this through a UEFI variable setting in the boot control device tree overlay. You'd have to look at the L4TLauncher code to track that down.

**Reply by Qbicz** [reply]
Even if you use extlinux, and your `extlinux.conf` point to a valid kernel device tree in `/boot` on your rootfs, the device tree from UEFI ("bootloader dtb") is still being passed and merged with the kernel device tree.

The bootloader dtb lives in `Linux_for_Tegra/bootloader`, but is being copied from `Linux_for_Tegra/kernel/dtb/` during build.
Also to make it more confusing, the bootloader device tree gets overlaid with e.g. dynamic dtbo, e.g. for Jetson Orin NX with `tegra234-p3768-0000+p3767-0000-dynamic.dtbo`.
At least this is the default behavior with BSP 36.4.

Knowing that you have to keep this bootloader dtb clean can save you from a lot of problems and confusion when changing the device tree.
Hope it helps!

**Reply by johannesschrimpf** [reply]
Thanks, that helps a lot. I was just wondering why the device tree still includes imx219-specific code even though I removed it from `tegra234-p3768-0000+p3767-0000-dynamic.dtbo`, but I only updated the rootfs (including my custom device tree using extlinux) via swupdate, not the boot loader.

It does feel like a but of a fragile setup with a kind-of dynamic picking of the dts, which either is from extlinux.conf, the boot partition, the boot loader itself and then in addition to be overlayed from the boot loader. In my case, where I use extlinux.conf now, a small mistake in my meta layer can result in a faulty extlinux.conf and transparent use of the dtb in the boot loader without me knowing.

We are using swupdate for updates for our customers in the field, so it's important that we can update the device tree if needed. I saw that there are "capsules" to update the boot loaders, but haven't looked much into it since the whole A/B partition setup is for robustness and it seems wrong to me to update the boot loaders in the field and risk something going wrong there.

**Reply by tbottomparallel** [reply]
@johannesschrimpf you may find that bootloader updates in the field are unavoidable - nvidia has placed a lot of firmware there (e.g. the sensor processing engine, realtime camera engine, and programmable vision accelerator firmware). If your bootloader version gets out of sync with userspace (you can check the bootloader l4t version with `nvbootctrl dump-slots-info`) it can break things. I ran into this recently - I updated just the bootloader to test a bugfix in the new UEFI version, and with r36.3.0 userspace and r36.4.0 bootloader cameras no longer worked because the driver and firmware were out of sync. AFAIK there is no way to update just the firmware from userspace, it's all part of the capsule blob.

A word of warning - UEFI capsule updates are one-way. You can't roll back to an earlier version without a hard flash <https://forums.developer.nvidia.com/t/orin-nano-r36-3-0-and-r36-4-0-stuck-at-bootloader/308786/5>.

**Reply by dwalkes** [reply]
> I saw that there are "capsules" to update the boot loaders, but haven't looked much into it since the whole A/B partition setup is for robustness and it seems wrong to me to update the boot loaders in the field and risk something going wrong there.

There's a new option in the swupdate reference implementation at https://github.com/OE4T/tegra-demo-distro/pull/320 which helps with this as you can avoid doing the bootloader update until/unless L4T versions change.  If you overwrite the versions used yourself you can be even more selective as to when you do the capsule update.
